File: backend/app/api/quality.py
"""
RAG 质量监控 API（设置页「RAG 质量监控」面板的数据源）.

    GET /quality/stats?window=session|24h|7d|30d|all

与 /badcases/stats 的分工
────────────────────────
/badcases/stats 回答"有多少坏答案等着人审"（队列视角）；
本端点回答"系统最近答得怎么样"（比率视角）：

    window=session  → 进程内计数器（本次运行，实时但重启归零）
    window=24h/7d/30d/all → quality_events 表聚合（重启不丢，可看趋势）

比率的分母为 0 时返回 null，同时返回 samples（各指标的样本量），
让前端能把"—"解释成"该窗口内还没有这类问答"，而不是假装有数据。
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Query

from app.db.user_models import User
from app.services.monitoring_service import metrics_snapshot, quality_stats
from app.services.permissions import require_permission
from app.utils.logging import get_logger

logger = get_logger(__name__)

router = APIRouter(tags=["Quality"])

# 时间窗 → 秒数；session / all 是两种特殊窗口
_WINDOWS: dict[str, float | None] = {
    "24h": 24 * 3600,
    "7d": 7 * 24 * 3600,
    "30d": 30 * 24 * 3600,
    "all": None,
}


@router.get(
    "/quality/stats",
    summary="RAG 质量比率（按时间窗聚合）",
    description=(
        "window=session 返回进程内实时计数；24h/7d/30d/all 从 quality_events "
        "表聚合，重启不丢历史。比率分母为 0 时为 null，samples 给出各指标样本量。"
    ),
)
async def get_quality_stats(
    user: Annotated[User, Depends(require_permission("audit.read"))],
    window: str = Query(default="24h", pattern="^(session|24h|7d|30d|all)$"),
) -> dict:
    if window == "session":
        snap = metrics_snapshot()
        return {
            "window": window,
            "source": "in_process",
            "total_queries": snap["counters"].get("rag.queries.total", 0),
            "ratios": snap["ratios"],
            "samples": {
                "evidence_gate": snap["counters"].get("evidence_gate.total", 0),
                "citations_checked": snap["counters"].get("citation.checked", 0),
                "citation_answers": snap["counters"].get("citation.total", 0),
                "queries": snap["counters"].get("rag.queries.total", 0),
            },
            "by_intent": {
                k.removeprefix("rag.intent."): v
                for k, v in snap["counters"].items()
                if k.startswith("rag.intent.")
            },
            "latency_ms": {
                stage: {
                    "avg": lat.get("avg_ms"),
                    "p95": None,  # 进程内只记 sum/max，无分位数
                    "max": lat.get("max_ms"),
                }
                for stage, lat in snap["latencies"].items()
            },
            "citation_failure_reasons": {
                "unsupported": snap["counters"].get("citation.unsupported", 0),
                "hallucinated": 0,  # 进程内未单独计数幻觉引用
                "misattributed": snap["counters"].get("citation.misattributed", 0),
                "number_mismatch": snap["counters"].get("citation.number_mismatch", 0),
                "date_mismatch": snap["counters"].get("citation.date_mismatch", 0),
            },
            "generated_at": snap["generated_at"],
        }

    stats = await quality_stats(_WINDOWS[window])
    return {"window": window, "source": "database", **stats}


__all__ = ["router"]

File: lib/api/quality.ts
import { apiFetch } from "@/lib/api/client";

/**
 * RAG 质量监控 API（/quality/stats）.
 *
 * window=session 走进程内计数器（实时、重启归零）；
 * 24h/7d/30d/all 走 quality_events 表聚合（重启不丢、可看趋势）。
 * 比率分母为 0 时为 null，samples 给出各指标样本量 —— 面板据此把 "—"
 * 解释成"该窗口内暂无这类问答"，而不是显示一个无意义的空值。
 */

export type QualityWindow = "session" | "24h" | "7d" | "30d" | "all";

export interface QualityRatios {
  evidence_refuse_rate: number | null;
  citation_pass_rate: number | null;
  citation_unsupported_rate: number | null;
  output_guard_change_rate: number | null;
  refusal_rate: number | null;
}

export interface LatencyStats {
  avg: number | null;
  p95: number | null;
  max: number | null;
}

/** 后端原始返回：session 窗口的 latency_ms 是分阶段结构，这里先原样接住。 */
interface QualityStatsRaw {
  window: QualityWindow;
  source: "in_process" | "database";
  total_queries: number;
  ratios: Partial<QualityRatios>;
  samples: {
    evidence_gate: number;
    citations_checked: number;
    citation_answers: number;
    queries: number;
  };
  by_intent: Record<string, number>;
  latency_ms:
    | LatencyStats
    | Record<string, { avg_ms?: number | null; avg?: number | null; p95?: number | null; max_ms?: number | null; max?: number | null }>;
  citation_failure_reasons: Record<string, number>;
  generated_at: string;
}

export interface QualityStats {
  window: QualityWindow;
  source: "in_process" | "database";
  totalQueries: number;
  ratios: Partial<QualityRatios>;
  samples: QualityStatsRaw["samples"];
  byIntent: Record<string, number>;
  /** 整体时延（已归一化：两种窗口结构统一为 avg/p95/max）。 */
  latency: LatencyStats;
  /** 分阶段时延（仅 session 窗口有）。 */
  stageLatencies: Record<string, LatencyStats>;
  citationFailureReasons: Record<string, number>;
  generatedAt: string;
}

function normalizeLatency(raw: QualityStatsRaw["latency_ms"]): {
  latency: LatencyStats;
  stageLatencies: Record<string, LatencyStats>;
} {
  const entries = Object.entries(raw ?? {});
  const isPerStage = entries.some(
    ([, v]) => v && typeof v === "object" && ("avg_ms" in v || "max_ms" in v)
  );
  if (!isPerStage) {
    // 数据库窗口：{ avg, p95, max }
    const flat = raw as LatencyStats;
    return {
      latency: { avg: flat.avg ?? null, p95: flat.p95 ?? null, max: flat.max ?? null },
      stageLatencies: {},
    };
  }
  // session 窗口：{ retrieval: {avg_ms, max_ms}, total: {...} }
  const stageLatencies: Record<string, LatencyStats> = {};
  for (const [stage, v] of entries) {
    if (!v || typeof v !== "object") continue;
    const rec = v as Record<string, number | null | undefined>;
    stageLatencies[stage] = {
      avg: rec.avg_ms ?? rec.avg ?? null,
      p95: rec.p95 ?? null,
      max: rec.max_ms ?? rec.max ?? null,
    };
  }
  return {
    latency: stageLatencies["total"] ?? { avg: null, p95: null, max: null },
    stageLatencies,
  };
}

/** 重置结果：scope 说明清的是哪一层，deleted 是实际删除的行数（session 窗口为 0）。 */
export interface QualityResetResult {
  ok: boolean;
  window: QualityWindow;
  scope: "in_process" | "database";
  deleted: number;
  resetAt: string;
}

/**
 * 重置质量统计（不可逆，界面必须二次确认）.
 *
 * 语义刻意与面板当前窗口对齐：session 只清进程内计数器（不动数据库）；
 * 24h/7d/30d/all 删除 quality_events 中该窗口内的行。
 */
export async function resetQualityStats(
  window: QualityWindow
): Promise<QualityResetResult> {
  const raw = await apiFetch<{
    ok: boolean;
    window: QualityWindow;
    scope: "in_process" | "database";
    deleted: number;
    reset_at: string;
  }>(`/quality/reset?window=${window}`, { method: "POST" });
  return {
    ok: raw.ok,
    window: raw.window,
    scope: raw.scope,
    deleted: raw.deleted ?? 0,
    resetAt: raw.reset_at ?? "",
  };
}

export async function getQualityStats(
  window: QualityWindow
): Promise<QualityStats> {
  const raw = await apiFetch<QualityStatsRaw>(`/quality/stats?window=${window}`);
  const { latency, stageLatencies } = normalizeLatency(raw.latency_ms);
  return {
    window: raw.window,
    source: raw.source,
    totalQueries: raw.total_queries ?? 0,
    ratios: raw.ratios ?? {},
    samples: raw.samples ?? {
      evidence_gate: 0,
      citations_checked: 0,
      citation_answers: 0,
      queries: 0,
    },
    byIntent: raw.by_intent ?? {},
    latency,
    stageLatencies,
    citationFailureReasons: raw.citation_failure_reasons ?? {},
    generatedAt: raw.generated_at ?? "",
  };
}
